### PR TITLE
Ddpb 1856 - professional deputies

### PR DESCRIPTION
**Pull request by elvisciotti**

### CODE REVIEW

File: app/config/security.yml [hunk 1]
```diff
@@ -93,15 +103,15 @@ security:
...
-        - { path: ^/deputyship-details, roles: [ROLE_DEPUTY, ROLE_ADMIN, ROLE_AD] }
+        - { path: ^/deputyship-details, roles: [ROLE_LAY_DEPUTY, ROLE_ADMIN, ROLE_AD] }
         - { path: ^/user-account/, roles: [ROLE_DEPUTY, ROLE_ADMIN, ROLE_AD] }
         - { path: ^/ad, roles: [ROLE_AD] }
-        - { path: ^/pa/team/add, roles: [ROLE_PA, ROLE_PA_ADMIN] }
-        - { path: ^/pa/team/edit, roles: [ROLE_PA, ROLE_PA_ADMIN, ROLE_PA_TEAM_MEMBER] }
-        - { path: ^/pa/team, roles: [ROLE_PA, ROLE_PA_ADMIN, ROLE_PA_TEAM_MEMBER] }
-        - { path: ^/pa/, roles: [ROLE_PA, ROLE_PA_ADMIN, ROLE_PA_TEAM_MEMBER] }
+        - { path: ^/pa/team/add, roles: [ROLE_PA_NAMED, ROLE_PA_ADMIN] }
+        - { path: ^/pa/team/edit, roles: [ROLE_PA] }
```
Comment (slizzio) [inline app/config/security.yml]: ROLE_PA_NAMED as well ?

File: app/config/security.yml [hunk 2]
```diff
@@ -93,15 +103,15 @@ security:
...
+        - { path: ^/pa/team, roles: [ROLE_PA] }
```
Comment (slizzio) [inline app/config/security.yml]: ROLE_PA_NAMED here too ?

File: src/AppBundle/Controller/Pa/TeamController.php [hunk 3]
```diff
@@ -48,12 +48,17 @@ public function addAction(Request $request)
         $form->handleRequest($request);
 
         if ($form->isValid()) {
+            /** @var $user EntityDir\User */
             $user = $form->getData();
 
-            if (!in_array($user->getRoleName(), [EntityDir\User::ROLE_PA_ADMIN, EntityDir\User::ROLE_PA_TEAM_MEMBER])) {
+            if ($this->isGranted(EntityDir\User::ROLE_PA) && !in_array($user->getRoleName(), [EntityDir\User::ROLE_PA_ADMIN, EntityDir\User::ROLE_PA_TEAM_MEMBER])) {
```
Comment (slizzio) [inline src/AppBundle/Controller/Pa/TeamController.php]: could this just be if ($this->isGranted(EntityDir\User::ROLE_PA_NAMED) ?

File: src/AppBundle/Controller/SettingsController.php [hunk 4]
```diff
@@ -93,22 +93,17 @@ public function profileEditAction(Request $request)
...
+            $form = $this->createForm(FormDir\User\UserDetailsBasicType::class, $user, []);
+            $jmsPutGroups = ['user_details_basic'];
+        } else if ($this->isGranted(EntityDir\User::ROLE_LAY_DEPUTY)) {
+            $form = $this->createForm(FormDir\Settings\ProfileType::class, $user, ['validation_groups' => ['user_details_full']]);
+            $jmsPutGroups = ['user_details_full'];
+        } else if ($this->isGranted(EntityDir\User::ROLE_PA)) {
+            $form = $this->createForm(FormDir\Settings\ProfileType::class, $user, ['validation_groups' => ['user_details_pa', 'profile_pa']]);
+            $jmsPutGroups = ['user_details_pa', 'profile_pa'];
+        } else {
+            $this->createAccessDeniedException('User role not recognised');
```
Comment (slizzio) [inline src/AppBundle/Controller/SettingsController.php]: switch statement is much easier on the eye :)

File: tests/behat/features/prof/03-report/01-102/01-report-edit-period.feature [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+#      | report_edit_endDate_year    | 2016 |
+#    # restore initial values (for future tests and have this test not affecting subsequent scenarios)
+#    And I load the application status from "prof-team-users-complete"
+#
+#  Scenario: PROF admin has access to edit 102 report dates
+#    Given I load the application status from "prof-team-users-complete"
+#    And I am logged in as "behat-prof1-team-member@publicguardian.gsi.gov.uk" with password "Abcd1234"
+#    When I click on "pa-report-open" in the "client-01000014" region
+#    And I click on "edit-report-period"
+#    Then the response status code should be 200
```
Comment (slizzio) [inline tests/behat/features/prof/03-report/01-102/01-report-edit-period.feature]: this file will need its own dedicated folder. Hopefully we can reuse some of the common components from the various report types

File: tests/behat/features/prof/03-report/01-102/02-common-report-sections.feature [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+#    And I click on "delete-documents-button" in the "document-list" region
+#    Then the response status code should be 200
+#    # delete this time
+#    And I click on "document-delete"
+#    Then the form should be valid
+#    And the URL should match "/report/\d+/documents/step/1"
+#    # Check document removed
+#    And the step with the following values CAN be submitted:
+#      | document_wishToProvideDocumentation_0 | no |
+#    And I save the application status into "102-common-sections-complete"
```
Comment (slizzio) [inline tests/behat/features/prof/03-report/01-102/02-common-report-sections.feature]: Same here

File: tests/behat/features/prof/03-report/01-102/03-edit-report-sections.feature [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+#      | account_amount      | 50.00     |
+#      # add another: no
+#    And I choose "no" when asked for adding another record
+#
+#
+#  Scenario: PROF 102 Report should be submittable
+#    Given I am logged in as "behat-prof1@publicguardian.gsi.gov.uk" with password "Abcd1234"
+#    And I click on "pa-report-open" in the "client-01000014" region
+#    Then the PROF report should be submittable
+#    And I save the application status into "prof-report-completed"
```
Comment (slizzio) [inline tests/behat/features/prof/03-report/01-102/03-edit-report-sections.feature]: and here maybe just put a big TODO at the top to notifty whoever enables it

File: app/config/security.yml (hunk 2 above)
Comment (elvisciotti) [reply app/config/security.yml]: need to fix team things. initially under ROLE_ORG

Comment (elvisciotti) [reply app/config/security.yml]: ROLE_PA now means ROLE_PA_*

File: src/AppBundle/Controller/SettingsController.php (hunk 4 above)
Comment (elvisciotti) [reply src/AppBundle/Controller/SettingsController.php]: not that easy, I'm using "isGranted" that uses the inheritance now. a switch will require all the sub-roles to be defined and will be more difficult to maintain

**Pull request merged**